File: 3329-find-the-length-of-the-longest-common-prefix/find-the-length-of-the-longest-common-prefix.py
class TrieNode:
    def __init__(self):
        self.child = [None]*10

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, num):
        node = self.root
        str_num = str(num)

        for digit in str_num:
            idx = int(digit)
            if node.child[idx] is None:
                node.child[idx] = TrieNode()
            node = node.child[idx]
    def find_longest_prefix(self,num):
        node = self.root
        str_num = str(num)
        length=0
        for digit in str_num:
            idx = int(digit)
            if (node.child[idx] is not None):
                length+=1
                node = node.child[idx]
            else:
                break
        return length

class Solution:
    def longestCommonPrefix(self, arr1: List[int], arr2: List[int]) -> int:
        trie = Trie()

        for num in arr1:
            trie.insert(num)

        ans = 0
        for num in arr2:
            ans = max(trie.find_longest_prefix(num), ans)

        return ans


"""
Brute-force solution:
-> O(mn)

Trie:
O(8m)+O(8n)
"""
        

File: 3329-find-the-length-of-the-longest-common-prefix/find-the-length-of-the-longest-common-prefix.cpp
class TrieNode{
public:
    TrieNode* child[10];
    TrieNode(){
        for (int i=0;i<10;++i){
            child[i]=NULL;
        }
    }
};

class Trie{
public:
    TrieNode* root;

    Trie() {root = new TrieNode();}

    void insert(int num){
        TrieNode* node = root;
        string numStr = to_string(num);

        for (auto c: numStr){
            if (node->child[c-'0']==NULL)
                node->child[c-'0'] = new TrieNode();
            node = node->child[c-'0'];
        }
    }

    int findLongestPrefix(int num){
        TrieNode* node = root;
        string numStr = to_string(num);
        int length=0;
        for(auto c:numStr){
            if (node->child[c-'0']!=NULL){
                length++;
                node = node->child[c-'0'];
            }
            else{
                break;
            }
        }
        return length;
    }
};


class Solution {
public:
    int longestCommonPrefix(vector<int>& arr1, vector<int>& arr2) {

        Trie trie;
        for (int num:arr1){
            trie.insert(num);
        }

        int ans=0;

        for(int num:arr2){
            ans = max(trie.findLongestPrefix(num), ans);
        }

        return ans;
        
    }
};